// coc-tsserver/lib/server/utils/logger.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const coc_nvim_1 = require("coc.nvim");
const is = tslib_1.__importStar(require("./is"));
class Logger {
    get output() {
        if (this._channel) {
            return this._channel;
        }
        this._channel = coc_nvim_1.workspace.createOutputChannel('tsserver');
        return this._channel;
    }
    dispose() {
        if (this._channel) {
            this._channel.dispose();
        }
    }
    data2String(data) {
        if (data instanceof Error) {
            if (is.string(data.stack)) {
                return data.stack;
            }
            return data.message;
        }
        if (is.boolean(data.success) && !data.success && is.string(data.message)) {
            return data.message;
        }
        if (is.string(data)) {
            return data;
        }
        return data.toString();
    }
    info(message, data) {
        this.logLevel('Info', message, data);
    }
    warn(message, data) {
        this.logLevel('Warn', message, data);
    }
    error(message, data) {
        // See https://github.com/Microsoft/TypeScript/issues/10496
        if (data && data.message === 'No content available.') {
            return;
        }
        this.logLevel('Error', message, data);
    }
    logLevel(level, message, data) {
        this.output.appendLine(`[${level}  - ${new Date().toLocaleTimeString()}] ${message}`);
        if (data) {
            this.output.appendLine(this.data2String(data));
        }
    }
}
exports.default = Logger;
