// coc-tsserver/lib/server/utils/semicolon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fast_diff_1 = tslib_1.__importDefault(require("fast-diff"));
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
function removeSemicolon(document, edits) {
    let orig = document.getText();
    let content = vscode_languageserver_protocol_1.TextDocument.applyEdits(document, edits);
    let result = content.split('\n').map(s => s.replace(/;$/, '')).join('\n');
    if (result == content)
        return edits;
    let change = getChange(orig, result);
    return [{
            range: {
                start: document.positionAt(change.start),
                end: document.positionAt(change.end)
            },
            newText: change.newText
        }];
}
exports.removeSemicolon = removeSemicolon;
function getChange(oldStr, newStr) {
    let result = fast_diff_1.default(oldStr, newStr, 1);
    let curr = 0;
    let start = -1;
    let end = -1;
    let newText = '';
    let remain = '';
    for (let item of result) {
        let [t, str] = item;
        // equal
        if (t == 0) {
            curr = curr + str.length;
            if (start != -1)
                remain = remain + str;
        }
        else {
            if (start == -1)
                start = curr;
            if (t == 1) {
                newText = newText + remain + str;
                end = curr;
            }
            else {
                newText = newText + remain;
                end = curr + str.length;
            }
            remain = '';
            if (t == -1)
                curr = curr + str.length;
        }
    }
    return { start, end, newText };
}
