// coc-tsserver/lib/server/features/definitionProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const typeConverters = tslib_1.__importStar(require("../utils/typeConverters"));
class TypeScriptDefinitionProvider {
    constructor(client) {
        this.client = client;
    }
    getSymbolLocations(definitionType, document, position, token) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const filepath = this.client.toPath(document.uri);
            if (!filepath) {
                return undefined;
            }
            const args = typeConverters.Position.toFileLocationRequestArgs(filepath, position);
            try {
                const response = yield this.client.execute(definitionType, args, token);
                const locations = (response && response.body) || [];
                return locations.map(location => typeConverters.Location.fromTextSpan(this.client.toResource(location.file), location));
            }
            catch (_a) {
                return [];
            }
        });
    }
    provideDefinition(document, position, token) {
        return this.getSymbolLocations('definition', document, position, token);
    }
    provideTypeDefinition(document, position, token) {
        return this.getSymbolLocations('typeDefinition', document, position, token);
    }
    provideImplementation(document, position, token) {
        return this.getSymbolLocations('implementation', document, position, token);
    }
}
exports.default = TypeScriptDefinitionProvider;
