// coc-tsserver/lib/server/features/hover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const previewer_1 = require("../utils/previewer");
const typeConverters = tslib_1.__importStar(require("../utils/typeConverters"));
class TypeScriptHoverProvider {
    constructor(client) {
        this.client = client;
    }
    provideHover(document, position, token) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const filepath = this.client.toPath(document.uri);
            if (!filepath) {
                return undefined;
            }
            const args = typeConverters.Position.toFileLocationRequestArgs(filepath, position);
            try {
                const response = yield this.client.execute('quickinfo', args, token);
                if (response && response.body) {
                    const data = response.body;
                    return {
                        contents: TypeScriptHoverProvider.getContents(data),
                        range: typeConverters.Range.fromTextSpan(data)
                    };
                }
            }
            catch (e) {
                // noop
            }
            return undefined;
        });
    }
    static getContents(data) {
        const parts = [];
        if (data.displayString) {
            parts.push({ language: 'typescript', value: data.displayString });
        }
        const tags = previewer_1.tagsMarkdownPreview(data.tags);
        parts.push(data.documentation + (tags ? '\n\n' + tags : ''));
        return parts;
    }
}
exports.default = TypeScriptHoverProvider;
